src/assign_images.py

```diff
@@ -0,0 +1,104 @@
+from loguru import logger as log
+import numpy as np
+from pandas import Series
+from requests.exceptions import HTTPError
+from tenacity import RetryError
+from tqdm import tqdm
+from typer import Argument, Option, Typer
+
+from src.images.image_source import ImageSourceSelector
+from src.images.local_images import LocalImages
+from src.images.mapillary import Mapillary
+
+app = Typer()
+
+
+@app.command()
+def main(
+    points_file: Annotated[
+        Path,
+        Argument(help="Path to Input Points File"),
+    ],
+    image_source: Annotated[
+        ImageSourceSelector, Argument(help="Where to Get Images From")
+    ],
+    images_path: Annotated[
+        Path,
+        Argument(help="Where the Images Should Be Located"),
+    ],
+    verbose: Annotated[bool, Option(help="Sets Log Level to DEBUG")] = False,
+):
+    """
+    Assigns Images to Points
+    :param points_file: Path to Input Points File.
+    File Format should be readable by :func:`geopandas.read_file`
+    :param image_source: Where to Get Images From.
+    :param images_path: Where the Images Should Be Located.
+    :param verbose: Sets Log Level to DEBUG
+    :return: None
+    """
+
+    log.remove()
+    if verbose:
+        log.add(sys.stdout, level="DEBUG")
+    else:
+        log.add(sys.stdout, level="INFO")
+
+    source = None
+    if image_source == ImageSourceSelector.local:
+        source = LocalImages(images_path)
+    elif image_source == ImageSourceSelector.mapillary:
+        source = Mapillary(getenv("MAPILLARY_CLIENT_TOKEN"), images_path)
+    else:
+        raise ValueError(f"Unknown Image Source: {image_source}")
+
+    gdf = gpd.read_file(points_file)
+    gdf["image_id"] = Series()
+    gdf["image_lat"] = Series()
+    gdf["image_lon"] = Series()
+    gdf["residual"] = Series()
+    gdf["image_path"] = Series()
+    gdf["error"] = Series()
+
+    for i, point in tqdm(
+        gdf.iterrows(),
+        total=len(gdf.index),
+        desc="Assigning Images to Points",
+        unit="points",
+    ):
+        latitude = point["geometry"].y
+        longitude = point["geometry"].x
+
+        try:
+            results = source.get_image_from_coordinates(latitude, longitude)
+            gdf.at[i, "image_lat"] = results["image_lat"]
+            gdf.at[i, "image_lon"] = results["image_lon"]
+            gdf.at[i, "residual"] = results["residual"]
+            gdf.at[i, "image_id"] = results["image_id"]
+            gdf.at[i, "image_path"] = str(results["image_path"])
+            gdf.at[i, "error"] = results["error"]
+        except HTTPError or RetryError as e:
+            log.error(e)
+            gdf.at[i, "error"] = e.__class__.__name__
+
+    log.info(gdf.head())
+    log.info(
+        "Are There Duplicates? {}",
+        gdf[gdf["image_id"] is not None and gdf["image_id"] != np.NaN]["image_id"]
+        .duplicated()
+        .any(),
+    )
+
+    gdf.to_file(
+        Path(points_file.parent, f"{points_file.stem}_images.gpkg"), driver="GPKG"
```

Comment:
Would be nice to log out the path that the new file is written out to.

loguru has a `log.success` level you can use for this.

Reply:
Output file log added in latest rev